File: main/linux/device_analyzer.cpp
#include "device_analyzer.h"
#include <fstream>
#include <sstream>
#include <iostream>
#include <algorithm>
#include <filesystem>
#include <cstdio>

namespace fs = std::filesystem;

// ==================== USB Устройства ====================
std::vector<DeviceAnalyzer::USBDeviceInfo> DeviceAnalyzer::get_usb_devices() {
    std::vector<USBDeviceInfo> devices;
    try {
        std::string usb_path = "/sys/bus/usb/devices/";
        for (const auto& entry : fs::directory_iterator(usb_path)) {
            std::string path = entry.path().string();
            std::string filename = entry.path().filename().string();
            if (filename.find(":") == std::string::npos) continue;
            
            USBDeviceInfo info;
            info.bus = filename.substr(0, filename.find(":"));
            info.device = filename.substr(filename.find(":") + 1);
            info.power_ma = 0;
            info.status = "connected";
            
            std::ifstream vendor_file(path + "/manufacturer");
            if (vendor_file.is_open()) { std::getline(vendor_file, info.vendor); vendor_file.close(); }
            
            std::ifstream product_file(path + "/product");
            if (product_file.is_open()) { std::getline(product_file, info.product); product_file.close(); }
            
            std::ifstream serial_file(path + "/serial");
            if (serial_file.is_open()) { std::getline(serial_file, info.serial); serial_file.close(); }
            
            std::ifstream power_file(path + "/bMaxPower");
            if (power_file.is_open()) { 
                std::string power_str; std::getline(power_file, power_str); 
                info.power_ma = std::stoi(power_str); power_file.close(); 
            }
            
            if (!info.vendor.empty() || !info.product.empty()) devices.push_back(info);
        }
    } catch (const std::exception& e) { std::cerr << "Error: " << e.what() << std::endl; }
    return devices;
}

std::string DeviceAnalyzer::get_usb_device_type(const USBDeviceInfo& device) {
    std::string product_lower = device.product;
    std::transform(product_lower.begin(), product_lower.end(), product_lower.begin(), ::tolower);
    
    if (product_lower.find("flash") != std::string::npos || product_lower.find("storage") != std::string::npos) return "flash_drive";
    if (product_lower.find("headphone") != std::string::npos || product_lower.find("audio") != std::string::npos) return "audio_device";
    if (product_lower.find("keyboard") != std::string::npos) return "keyboard";
    if (product_lower.find("mouse") != std::string::npos) return "mouse";
    if (product_lower.find("camera") != std::string::npos || product_lower.find("webcam") != std::string::npos) return "camera";
    if (product_lower.find("ethernet") != std::string::npos || product_lower.find("network") != std::string::npos) return "network_adapter";
    return "unknown";
}

// ==================== Батарея и зарядка ====================
DeviceAnalyzer::BatteryInfo DeviceAnalyzer::get_battery_info() {
    BatteryInfo info;
    info.is_present = false;
    info.is_charging = false;
    info.percentage = 0;
    info.voltage_mv = 0;
    info.current_ma = 0;
    info.power_mw = 0;
    
    try {
        std::string battery_path = "/sys/class/power_supply/";
        for (const auto& entry : fs::directory_iterator(battery_path)) {
            std::string path = entry.path().string();
            std::string name = entry.path().filename().string();
            if (name.find("BAT") != 0) continue;
            
            info.is_present = true;
            
            std::ifstream status_file(path + "/status");
            if (status_file.is_open()) { 
                std::getline(status_file, info.status);
                info.is_charging = (info.status == "Charging");
                status_file.close();
            }
            
            std::ifstream capacity_file(path + "/capacity");
            if (capacity_file.is_open()) { 
                std::string cap; std::getline(capacity_file, cap);
                info.percentage = std::stoi(cap); capacity_file.close();
            }
            
            std::ifstream voltage_file(path + "/voltage_now");
            if (voltage_file.is_open()) { 
                std::string volt; std::getline(voltage_file, volt);
                info.voltage_mv = std::stoi(volt) / 1000; voltage_file.close();
            }
            
            std::ifstream current_file(path + "/current_now");
            if (current_file.is_open()) { 
                std::string curr; std::getline(current_file, curr);
                info.current_ma = std::stoi(curr) / 1000; current_file.close();
            }
            
            if (info.voltage_mv > 0 && info.current_ma > 0) {
                info.power_mw = (info.voltage_mv * info.current_ma) / 1000;
            }
            
            std::ifstream manufacturer_file(path + "/manufacturer");
            if (manufacturer_file.is_open()) { std::getline(manufacturer_file, info.manufacturer); manufacturer_file.close(); }
            
            std::ifstream model_file(path + "/model_name");
            if (model_file.is_open()) { std::getline(model_file, info.model); model_file.close(); }
            
            std::ifstream tech_file(path + "/technology");
            if (tech_file.is_open()) { std::getline(tech_file, info.technology); tech_file.close(); }
            
            break;
        }
    } catch (const std::exception& e) { std::cerr << "Error: " << e.what() << std::endl; }
    return info;
}

std::vector<std::string> DeviceAnalyzer::detect_battery_issues(const BatteryInfo& info) {
    std::vector<std::string> issues;
    if (!info.is_present) { issues.push_back("Батарея не обнаружена"); return issues; }
    if (info.percentage < 20 && !info.is_charging) issues.push_back("Низкий заряд батареи (< 20%)");
    if (info.voltage_mv > 0 && info.voltage_mv < 10000) issues.push_back("Низкое напряжение (возможен износ)");
    if (info.current_ma > 0 && info.is_charging && info.current_ma < 500) issues.push_back("Медленная зарядка (< 500 мА)");
    if (info.technology.empty()) issues.push_back("Не удалось определить технологию батареи");
    return issues;
}

// ==================== Аудио устройства ====================
std::vector<DeviceAnalyzer::AudioDeviceInfo> DeviceAnalyzer::get_audio_devices() {
    std::vector<AudioDeviceInfo> devices;
    try {
        FILE* pipe = popen("aplay -l", "r");
        if (pipe) {
            char buffer[256];
            while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
                std::string line(buffer);
                if (line.find("card") != std::string::npos) {
                    AudioDeviceInfo info;
                    size_t card_pos = line.find("card ");
                    if (card_pos != std::string::npos) {
                        size_t colon_pos = line.find(":", card_pos);
                        info.card = line.substr(card_pos + 5, colon_pos - card_pos - 5);
                        size_t device_pos = line.find("device ");
                        if (device_pos != std::string::npos) {
                            size_t colon2_pos = line.find(":", device_pos);
                            info.device = line.substr(device_pos + 7, colon2_pos - device_pos - 7);
                        }
                        size_t colon3_pos = line.find(": ", colon_pos + 2);
                        if (colon3_pos != std::string::npos) {
                            info.name = line.substr(colon3_pos + 2);
                            if (!info.name.empty() && info.name.back() == '\n') info.name.pop_back();
                        }
                        
                        std::string name_lower = info.name;
                        std::transform(name_lower.begin(), name_lower.end(), name_lower.begin(), ::tolower);
                        if (name_lower.find("headphone") != std::string::npos || name_lower.find("headset") != std::string::npos) info.type = "headphones";
                        else if (name_lower.find("microphone") != std::string::npos) info.type = "microphone";
                        else if (name_lower.find("speaker") != std::string::npos) info.type = "speakers";
                        else info.type = "unknown";
                        info.status = "connected";
                        devices.push_back(info);
                    }
                }
            }
            pclose(pipe);
        }
    } catch (const std::exception& e) { std::cerr << "Error: " << e.what() << std::endl; }
    return devices;
}

std::vector<std::string> DeviceAnalyzer::detect_audio_issues(const AudioDeviceInfo& device) {
    std::vector<std::string> issues;
    if (device.name.empty()) issues.push_back("Не удалось определить имя устройства");
    if (device.type == "unknown") issues.push_back("Неизвестный тип аудиоустройства");
    return issues;
}

// ==================== Ввод (клавиатура, мышь) ====================
std::vector<DeviceAnalyzer::InputDeviceInfo> DeviceAnalyzer::get_input_devices() {
    std::vector<InputDeviceInfo> devices;
    try {
        std::ifstream input_file("/proc/bus/input/devices");
        if (input_file.is_open()) {
            std::string line;
            InputDeviceInfo current_device;
            bool in_device = false;
            while (std::getline(input_file, line)) {
                if (line.empty()) {
                    if (in_device && !current_device.name.empty()) {
                        current_device.is_connected = true;
                        current_device.status = "connected";
                        devices.push_back(current_device);
                    }
                    current_device = InputDeviceInfo();
                    in_device = false;
                    continue;
                }
                if (line.find("N: Name=") == 0) {
                    current_device.name = line.substr(8);
                    if (current_device.name.front() == '"' && current_device.name.back() == '"') {
                        current_device.name = current_device.name.substr(1, current_device.name.length() - 2);
                    }
                    in_device = true;
                    std::string name_lower = current_device.name;
                    std::transform(name_lower.begin(), name_lower.end(), name_lower.begin(), ::tolower);
                    if (name_lower.find("keyboard") != std::string::npos) current_device.type = "keyboard";
                    else if (name_lower.find("mouse") != std::string::npos) current_device.type = "mouse";
                    else if (name_lower.find("touchpad") != std::string::npos) current_device.type = "touchpad";
                    else current_device.type = "unknown";
                }
                if (line.find("B: BUS=") == 0) current_device.bus = line.substr(7);
                if (line.find("P: Vendor=") == 0) current_device.vendor = line.substr(11);
                if (line.find("P: Product=") == 0) current_device.product = line.substr(12);
            }
            if (in_device && !current_device.name.empty()) {
                current_device.is_connected = true;
                current_device.status = "connected";
                devices.push_back(current_device);
            }
            input_file.close();
        }
    } catch (const std::exception& e) { std::cerr << "Error: " << e.what() << std::endl; }
    return devices;
}

std::vector<std::string> DeviceAnalyzer::detect_input_issues(const InputDeviceInfo& device) {
    std::vector<std::string> issues;
    if (device.type == "unknown") issues.push_back("Неизвестный тип устройства ввода");
    if (device.vendor.empty()) issues.push_back("Не удалось определить производителя");
    return issues;
}

// ==================== Гнезда ноутбука ====================
DeviceAnalyzer::PortInfo DeviceAnalyzer::get_port_info() {
    PortInfo info;
    info.is_laptop = false;
    info.has_hdmi = false;
    info.has_usb_c = false;
    info.has_usb_a = false;
    info.has_ethernet = false;
    info.has_audio_jack = false;
    info.has_sd_card = false;
    info.chassis_type = "unknown";
    
    try {
        std::ifstream dmi_file("/sys/class/dmi/id/chassis_type");
        if (dmi_file.is_open()) {
            std::string chassis_str;
            std::getline(dmi_file, chassis_str);
            int chassis_type = std::stoi(chassis_str);
            if (chassis_type >= 8 && chassis_type <= 10 || chassis_type == 14) {
                info.is_laptop = true;
                info.chassis_type = "laptop";
            } else if (chassis_type >= 3 && chassis_type <= 7) {
                info.chassis_type = "desktop";
            }
            dmi_file.close();
        }
        
        FILE* pipe = popen("lspci", "r");
        if (pipe) {
            char buffer[256];
            while (fgets(buffer, sizeof(buffer), pipe) != nullptr) {
                std::string line(buffer);
                std::transform(line.begin(), line.end(), line.begin(), ::tolower);
                if (line.find("hdmi") != std::string::npos) info.has_hdmi = true;
                if (line.find("ethernet") != std::string::npos || line.find("network") != std::string::npos) info.has_ethernet = true;
            }
            pclose(pipe);
        }
        
        for (const auto& entry : fs::directory_iterator("/sys/bus/usb/devices/")) {
            std::string filename = entry.path().filename().string();
            if (filename.find("usb") == 0) {
                info.has_usb_a = true;
                std::ifstream speed_file(entry.path().string() + "/speed");
                if (speed_file.is_open()) {
                    std::string speed; std::getline(speed_file, speed);
                    if (speed.find("10000") != std::string::npos || speed.find("5000") != std::string::npos) {
                        info.has_usb_c = true;
                    }
                    speed_file.close();
                }
            }
        }
        
        info.has_audio_jack = true; // По умолчанию для большинства систем
        
    } catch (const std::exception& e) { std::cerr << "Error: " << e.what() << std::endl; }
    return info;
}

File: main/linux/device_analyzer.py
"""
Device Analyzer - Анализатор подключенных устройств
"""

import subprocess
import os
from typing import List, Dict


class DeviceAnalyzer:
    """Анализатор подключенных устройств"""
    
    def __init__(self):
        pass
    
    def get_usb_devices(self) -> List[Dict[str, str]]:
        """Получить информацию о USB устройствах"""
        devices = []
        
        try:
            result = subprocess.run(['lsusb'], capture_output=True, text=True, timeout=5)
            lines = result.stdout.split('\n')
            
            for line in lines:
                if line.strip():
                    parts = line.split()
                    if len(parts) >= 6:
                        device = {
                            'bus': parts[1],
                            'device': parts[3],
                            'id': parts[5],
                            'description': ' '.join(parts[6:]),
                            'type': 'USB'
                        }
                        devices.append(device)
        except Exception as e:
            print(f"Ошибка при получении USB устройств: {e}")
        
        return devices
    
    def get_pci_devices(self) -> List[Dict[str, str]]:
        """Получить информацию о PCI устройствах"""
        devices = []
        
        try:
            result = subprocess.run(['lspci', '-v'], capture_output=True, text=True, timeout=5)
            lines = result.stdout.split('\n')
            
            current_device = None
            for line in lines:
                if ':' in line and not line.startswith('\t'):
                    if current_device:
                        devices.append(current_device)
                    
                    parts = line.split(':', 1)
                    current_device = {
                        'slot': parts[0].strip(),
                        'description': parts[1].strip(),
                        'details': []
                    }
                elif current_device and line.strip():
                    if ':' in line:
                        key, value = line.split(':', 1)
                        current_device['details'].append({
                            'key': key.strip(),
                            'value': value.strip()
                        })
            
            if current_device:
                devices.append(current_device)
                
        except Exception as e:
            print(f"Ошибка при получении PCI устройств: {e}")
        
        return devices
    
    def get_input_devices(self) -> List[Dict[str, str]]:
        """Получить информацию об устройствах ввода"""
        devices = []
        
        # Проверяем /dev/input
        input_dir = '/dev/input'
        if os.path.exists(input_dir):
            try:
                for item in os.listdir(input_dir):
                    if item.startswith('event'):
                        devices.append({
                            'name': item,
                            'path': os.path.join(input_dir, item),
                            'type': 'Input Device'
                        })
            except PermissionError:
                pass
        
        # Проверяем через xinput если доступно
        try:
            result = subprocess.run(['xinput', '--list'], capture_output=True, text=True, timeout=5)
            lines = result.stdout.split('\n')
            
            for line in lines:
                if '⎡' in line or '↳' in line:
                    parts = line.split()
                    if len(parts) >= 2:
                        device_id = parts[-1].strip('[]')
                        device_name = ' '.join(parts[1:-1])
                        devices.append({
                            'name': device_name,
                            'id': device_id,
                            'type': 'Input Device'
                        })
        except Exception:
            pass
        
        return devices
    
    def get_all_devices(self) -> Dict[str, List[Dict[str, str]]]:
        """Получить все устройства"""
        return {
            'usb': self.get_usb_devices(),
            'pci': self.get_pci_devices(),
            'input': self.get_input_devices()
        }
    
    def get_device_summary(self) -> Dict[str, int]:
        """Получить сводку по устройствам"""
        all_devices = self.get_all_devices()
        
        return {
            'usb_count': len(all_devices['usb']),
            'pci_count': len(all_devices['pci']),
            'input_count': len(all_devices['input']),
            'total': len(all_devices['usb']) + len(all_devices['pci']) + len(all_devices['input'])
        }
    
    def get_battery_info(self) -> Dict[str, any]:
        """Получить информацию о батарее и зарядке"""
        info = {
            'is_present': False,
            'is_charging': False,
            'percentage': 0,
            'voltage_mv': 0,
            'current_ma': 0,
            'power_mw': 0,
            'manufacturer': '',
            'model': '',
            'technology': '',
            'status': ''
        }
        
        try:
            battery_path = '/sys/class/power_supply/'
            if os.path.exists(battery_path):
                for item in os.listdir(battery_path):
                    if item.startswith('BAT'):
                        bat_dir = os.path.join(battery_path, item)
                        info['is_present'] = True
                        
                        # Читаем статус
                        status_file = os.path.join(bat_dir, 'status')
                        if os.path.exists(status_file):
                            with open(status_file, 'r') as f:
                                info['status'] = f.read().strip()
                                info['is_charging'] = info['status'] == 'Charging'
                        
                        # Читаем процент
                        capacity_file = os.path.join(bat_dir, 'capacity')
                        if os.path.exists(capacity_file):
                            with open(capacity_file, 'r') as f:
                                info['percentage'] = int(f.read().strip())
                        
                        # Читаем напряжение
                        voltage_file = os.path.join(bat_dir, 'voltage_now')
                        if os.path.exists(voltage_file):
                            with open(voltage_file, 'r') as f:
                                info['voltage_mv'] = int(f.read().strip()) // 1000
                        
                        # Читаем ток
                        current_file = os.path.join(bat_dir, 'current_now')
                        if os.path.exists(current_file):
                            with open(current_file, 'r') as f:
                                info['current_ma'] = int(f.read().strip()) // 1000
                        
                        # Вычисляем мощность
                        if info['voltage_mv'] > 0 and info['current_ma'] > 0:
                            info['power_mw'] = (info['voltage_mv'] * info['current_ma']) // 1000
                        
                        # Читаем производителя
                        manufacturer_file = os.path.join(bat_dir, 'manufacturer')
                        if os.path.exists(manufacturer_file):
                            with open(manufacturer_file, 'r') as f:
                                info['manufacturer'] = f.read().strip()
                        
                        # Читаем модель
                        model_file = os.path.join(bat_dir, 'model_name')
                        if os.path.exists(model_file):
                            with open(model_file, 'r') as f:
                                info['model'] = f.read().strip()
                        
                        # Читаем технологию
                        tech_file = os.path.join(bat_dir, 'technology')
                        if os.path.exists(tech_file):
                            with open(tech_file, 'r') as f:
                                info['technology'] = f.read().strip()
                        
                        break
        except Exception as e:
            print(f"Ошибка при получении информации о батарее: {e}")
        
        return info
    
    def detect_battery_issues(self, battery_info: Dict[str, any]) -> List[str]:
        """Обнаружить проблемы с батареей"""
        issues = []
        
        if not battery_info['is_present']:
            issues.append('Батарея не обнаружена')
            return issues
        
        if battery_info['percentage'] < 20 and not battery_info['is_charging']:
            issues.append('Низкий заряд батареи (< 20%)')
        
        if battery_info['voltage_mv'] > 0 and battery_info['voltage_mv'] < 10000:
            issues.append('Низкое напряжение батареи (возможен износ)')
        
        if battery_info['current_ma'] > 0 and battery_info['is_charging'] and battery_info['current_ma'] < 500:
            issues.append('Медленная зарядка (ток < 500 мА)')
        
        if not battery_info['technology']:
            issues.append('Не удалось определить технологию батареи')
        
        return issues
    
    def get_audio_devices(self) -> List[Dict[str, str]]:
        """Получить информацию об аудиоустройствах"""
        devices = []
        
        try:
            result = subprocess.run(['aplay', '-l'], capture_output=True, text=True, timeout=5)
            lines = result.stdout.split('\n')
            
            current_device = None
            for line in lines:
                if 'card' in line:
                    if current_device:
                        devices.append(current_device)
                    
                    parts = line.split()
                    current_device = {'name': '', 'type': 'unknown', 'status': 'connected'}
                    
                    # Парсим номер карты
                    if 'card' in parts:
                        card_idx = parts.index('card')
                        if card_idx + 1 < len(parts):
                            current_device['card'] = parts[card_idx + 1].rstrip(':')
                    
                    # Парсим номер устройства
                    if 'device' in parts:
                        device_idx = parts.index('device')
                        if device_idx + 1 < len(parts):
                            current_device['device'] = parts[device_idx + 1].rstrip(':')
                    
                    # Получаем имя устройства
                    if ':' in line:
                        name = line.split(':', 1)[1].strip()
                        current_device['name'] = name
                        
                        # Определяем тип
                        name_lower = name.lower()
                        if 'headphone' in name_lower or 'headset' in name_lower:
                            current_device['type'] = 'headphones'
                        elif 'microphone' in name_lower:
                            current_device['type'] = 'microphone'
                        elif 'speaker' in name_lower:
                            current_device['type'] = 'speakers'
            
            if current_device:
                devices.append(current_device)
                
        except Exception as e:
            print(f"Ошибка при получении аудиоустройств: {e}")
        
        return devices
    
    def detect_audio_issues(self, audio_device: Dict[str, str]) -> List[str]:
        """Обнаружить проблемы с аудиоустройством"""
        issues = []
        
        if not audio_device.get('name'):
            issues.append('Не удалось определить имя устройства')
        
        if audio_device.get('type') == 'unknown':
            issues.append('Неизвестный тип аудиоустройства')
        
        return issues
    
    def get_port_info(self) -> Dict[str, any]:
        """Получить информацию о портах и типе устройства"""
        info = {
            'is_laptop': False,
            'has_hdmi': False,
            'has_usb_c': False,
            'has_usb_a': False,
            'has_ethernet': False,
            'has_audio_jack': False,
            'has_sd_card': False,
            'chassis_type': 'unknown'
        }
        
        try:
            # Проверяем DMI информацию
            dmi_file = '/sys/class/dmi/id/chassis_type'
            if os.path.exists(dmi_file):
                with open(dmi_file, 'r') as f:
                    chassis_type = int(f.read().strip())
                    # Типы корпусов: 8=Notebook, 9=Laptop, 10=Tablet, 14=Sub Notebook
                    if chassis_type in [8, 9, 10, 14]:
                        info['is_laptop'] = True
                        info['chassis_type'] = 'laptop'
                    elif chassis_type in [3, 4, 5, 6, 7]:
                        info['chassis_type'] = 'desktop'
            
            # Проверяем наличие портов через lspci
            result = subprocess.run(['lspci'], capture_output=True, text=True, timeout=5)
            lspci_output = result.stdout.lower()
            
            if 'hdmi' in lspci_output:
                info['has_hdmi'] = True
            if 'ethernet' in lspci_output or 'network' in lspci_output:
                info['has_ethernet'] = True
            
            # Проверяем USB порты
            usb_path = '/sys/bus/usb/devices/'
            if os.path.exists(usb_path):
                for item in os.listdir(usb_path):
                    if item.startswith('usb'):
                        info['has_usb_a'] = True
                        speed_file = os.path.join(usb_path, item, 'speed')
                        if os.path.exists(speed_file):
                            with open(speed_file, 'r') as f:
                                speed = f.read().strip()
                                if '10000' in speed or '5000' in speed:
                                    info['has_usb_c'] = True
            
            info['has_audio_jack'] = True  # По умолчанию для большинства систем
            
        except Exception as e:
            print(f"Ошибка при получении информации о портах: {e}")
        
        return info
    
    def get_usb_device_type(self, device_description: str) -> str:
        """Определить тип USB устройства по описанию"""
        desc_lower = device_description.lower()
        
        if 'flash' in desc_lower or 'storage' in desc_lower or 'disk' in desc_lower:
            return 'flash_drive'
        elif 'headphone' in desc_lower or 'audio' in desc_lower:
            return 'audio_device'
        elif 'keyboard' in desc_lower:
            return 'keyboard'
        elif 'mouse' in desc_lower:
            return 'mouse'
        elif 'camera' in desc_lower or 'webcam' in desc_lower:
            return 'camera'
        elif 'ethernet' in desc_lower or 'network' in desc_lower:
            return 'network_adapter'
        
        return 'unknown'
